### PR TITLE
Convert managedclusters service to SDKv2

### LINKED ISSUE #3861: Update managedclusters service to SDK v2
This also includes the agentpools service; the two aren't easily separated.

### PR DESCRIPTION
**Pull request by mboersma**
**What type of PR is this?**

/kind cleanup

**What this PR does / why we need it**:

Converts the managedclusters service and the agentpools service to use SDKv2 and the `asyncpoller` framework for long-running operations.

**Which issue(s) this PR fixes**:

Fixes #3861
Refs #3409

**Special notes for your reviewer**:

- [ ] cherry-pick candidate

**TODOs**:

- [x] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Review by CecileRobertMichon — COMMENTED** [review]
/lgtm
/hold for squash

**Comment by maciaszczykm** [comment]
@mboersma @CecileRobertMichon Is there anything that I can do to help get it merged? I would really like to be able to use https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/2498 and this one seems to be required before.

**Comment by mboersma** [comment]
@maciaszczykm thanks for wanting to help! This branch is working well in my local testing with `tilt`, and is provisioning AKS clusters here in e2e, but obviously failing the overall test. I'm just starting to look into that, but it seems like we're close to having this be mergeable.

If you're comfortable working with this branch locally, more testing always helps. And additional reviews are always welcome. Other than that, please feel free to look at the e2e failure logs and comment here if you see the problem.

**Comment by mboersma** [comment]
I ran the e2e tests locally and see the same result, so that's good: we can focus on this failure:

```
> By("creating a machine pool with public IP addresses from a prefix", func() {
```

```go
  STEP: Verifying the AzureManagedMachinePool converges to a failed ready status @ 08/24/23 22:22:47.771
  STEP: Scaling the MachinePool to 2 nodes @ 08/24/23 22:23:27.793
  STEP: Verifying the AzureManagedMachinePool becomes ready @ 08/24/23 22:23:27.811
  STEP: Deleting the node pool @ 08/24/23 22:48:27.812
  [FAILED] in [It] - /home/prow/go/src/sigs.k8s.io/cluster-api-provider-azure/test/e2e/aks_public_ip_prefix.go:171 @ 08/24/23 22:48:37.837
  Aug 24 22:48:37.837: INFO: FAILED!
  Aug 24 22:48:37.837: INFO: Cleaning up after "Workload cluster creation Creating an AKS cluster [Managed Kubernetes] with a single control plane node and 1 node" spec

**Comment by CecileRobertMichon** [comment]
@mboersma I tried poking around the logs, here's the error I found that seems relevant:

```
E0824 22:23:32.448417       1 controller.go:324]  "msg"="Reconciler error" "error"="error creating AzureManagedMachinePool capz-e2e-xqfxcy/pool3: failed to reconcile machine pool pool3: failed to update resource capz-e2e-xqfxcy-aks/pool3 (service: agentpools): GET https://management.azure.com/subscriptions/===REDACTED===/providers/Microsoft.ContainerService/locations/canadacentral/operations/3c242383-3be3-4495-8a3a-a1d35d286181\n--------------------------------------------------------------------------------\nRESPONSE 200: 200 OK\nERROR CODE: ReconcileVMSSAgentPoolFailed\n--------------------------------------------------------------------------------\n{\n  \"name\": \"8323243c-e33b-9544-8a3a-a1d35d286181\",\n  \"status\": \"Failed\",\n  \"startTime\": \"2023-08-24T22:23:06.9240862Z\",\n  \"endTime\": \"2023-08-24T22:23:12.5070864Z\",\n  \"error\": {\n    \"code\": \"ReconcileVMSSAgentPoolFailed\",\n    \"message\": \"Code=\\\"PublicIpPrefixOutOfIpAddressesForVMScaleSet\\\" Message=\\\"IpPrefix /subscriptions/===REDACTED===/resourceGroups/capz-e2e-xqfxcy-aks/providers/Microsoft.Network/publicIPPrefixes/capz-e2e-xqfxcy-aks can provide 2 public ips at maximum, 0 of them are in use already. Current available public ip Count for this prefix is 2, which is smaller than required number of public ip count 3 for VMScaleSet on this prefix.\\\" Details=[]\"\n  }\n}\n--------------------------------------------------------------------------------\n" "AzureManagedMachinePool"={"name":"pool3","namespace":"capz-e2e-xqfxcy"} "controller"="azuremanagedmachinepool" "controllerGroup"="infrastructure.cluster.x-k8s.io" "controllerKind"="AzureManagedMachinePool" "name"="pool3" "namespace"="capz-e2e-xqfxcy" "reconcileID"="37aba0e7-3695-4c0b-8d09-597d95217c4c"

```

seems like this is a change in the new API?

**Comment by mboersma** [comment]
@CecileRobertMichon good eye! I removed that part of the spec in a separate commit. 

If that gets things to pass, I'll open an issue to investigate whether we should restore the "AMMP-converges-to-failed-ready" part of the test.

File: test/e2e/aks_public_ip_prefix.go [hunk 1]
```diff
@@ -82,7 +82,7 @@ func AKSPublicIPPrefixSpec(ctx context.Context, inputGetter func() AKSPublicIPPr
 		g.Expect(err).NotTo(HaveOccurred())
 	}, input.WaitIntervals...).Should(Succeed(), "failed to create public IP prefix")
 
-	By("Creating node pool with 3 nodes")
+	By("Creating node pool with 2 nodes")
```
Comment (CecileRobertMichon) [inline test/e2e/aks_public_ip_prefix.go:85]: you need to also change the replica count on line 108 for this to be true :)

Comment (mboersma) [reply test/e2e/aks_public_ip_prefix.go:85]: Corrigé, merci !

**Comment by mboersma** [comment]
The aks test finally passed after I fixed a pointer comparison. w00t! I'll squash that fix and create an issue to track the spec removal above. 🎉

**Comment by mboersma** [comment]
/hold cancel

I know this PR has 4 commits, but I think they're logically separate, and it may be useful to keep the test removal as its own commit. But let me know if anyone disagrees; happy to squash further.

**Comment by mboersma** [comment]
I created #3893 to track the part of the test spec that this removes.

**Review by CecileRobertMichon — COMMENTED** [review]
/lgtm
/assign @nojnhuh

**Comment by mboersma** [comment]
/retest

Looks like a Windows flake? This always passed previously.

File: azure/services/agentpools/spec.go [hunk 2]
```diff
@@ -269,24 +269,18 @@ func (s *AgentPoolSpec) Parameters(ctx context.Context, existing interface{}) (p
 			log.V(4).Info("no changes found between user-updated spec and existing spec")
 			return nil, nil
 		}
-		log.V(4).Info("found a diff between the desired spec and the existing agentpool", "difference", diff)
```
Comment (nojnhuh) [inline azure/services/agentpools/spec.go:272]: Was changing the verbosity here intentional? I don't have an opinion on what it should be, but this is the kind of thing I would definitely do to help debug and then forget to change back 😄

File: azure/services/managedclusters/spec_test.go [hunk 3]
```diff
@@ -69,7 +69,7 @@ func TestParameters(t *testing.T) {
 					"test-tag": "test-value",
 				},
 				Version:         "v1.22.0",
-				LoadBalancerSKU: "Standard",
+				LoadBalancerSKU: "standard",
```
Comment (nojnhuh) [inline azure/services/managedclusters/spec_test.go:72]: Do we need to do some conversion ourselves so as not to break existing users with clusters that already specify (capital) `Standard` or `Basic`? Or is this not actually case-sensitive? I see none of our CI templates specify anything for this field.

File: go.mod [hunk 4]
```diff
@@ -8,6 +8,7 @@ require (
 	github.com/Azure/azure-sdk-for-go/sdk/azcore v1.7.1
 	github.com/Azure/azure-sdk-for-go/sdk/azidentity v1.3.1
 	github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/compute/armcompute/v5 v5.1.0
+	github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/containerservice/armcontainerservice/v4 v4.1.0
```
Comment (nojnhuh) [inline go.mod:11]: Just realized with the way the new SDK does modules we'll basically have dependabot doing Azure API bumps for us :tada:

File: azure/services/agentpools/spec.go (hunk 2 above)
Comment (mboersma) [reply azure/services/agentpools/spec.go:272]: Yep, forgot to revert a debugging change. :-) Will fix.

File: azure/services/managedclusters/spec_test.go (hunk 3 above)
Comment (mboersma) [reply azure/services/managedclusters/spec_test.go:72]: AFAICT in reality the API accepts either `Standard` or `standard` (and so on), but the API now prefers all lower-case in its constants and tests.

File: azure/services/agentpools/spec.go (hunk 2 above)
Comment (mboersma) [reply azure/services/agentpools/spec.go:272]: Fixed.

**Review by CecileRobertMichon — COMMENTED** [review]
/lgtm
/approve

File: controllers/azuremanagedcontrolplane_reconciler.go [hunk 5]
```diff
@@ -49,14 +49,18 @@ func newAzureManagedControlPlaneReconciler(scope *scope.ManagedControlPlaneScope
 	if scope.UseLegacyGroups {
 		groupsService = groups.New(scope)
 	}
+	managedClustersService, err := managedclusters.New(scope)
+	if err != nil {
+		panic(err) // TODO: handle this error differently
```
[the remark this answers is not in the document]
Comment (CecileRobertMichon) [reply controllers/azuremanagedcontrolplane_reconciler.go:54]: the function this is called from (reconcileNormal in the controller) does return an error, so it should be possible to return the error back to it

[the remark this answers is not in the document]
Comment (mboersma) [reply controllers/azuremanagedcontrolplane_reconciler.go:54]: /hold

Yes, I think so, although "differently" probably requires some refactoring to bubble the error upward. Unless you think we can handle that as a follow-up issue.

[the remark this answers is not in the document]
Comment (mboersma) [reply controllers/azuremanagedcontrolplane_reconciler.go]: It was actually called from 3 `reconcile`-type funcs, but they were all easily amenable to returning the error. 👍🏻

**Review by CecileRobertMichon — COMMENTED** [review]
/hold cancel
/lgtm
/approve

**Comment by CecileRobertMichon** [comment]
/test pull-cluster-api-provider-azure-windows-containerd-upstream-with-ci-artifacts

flake

**Comment by mboersma** [comment]
/hold

The AKS is test is taking an unusually long time, so even if it passes I'd like to verify that it's an outlier.

**Comment by mboersma** [comment]
/test pull-cluster-api-provider-azure-e2e-aks 

Let's see another green run.

**Comment by CecileRobertMichon** [comment]
@mboersma I checked the two previous e2e failures, they happened at two different places in the test so I don't think there's a pattern

**Pull request merged**